### PR TITLE
Add code review guidelines for changing running infrastructure

### PR DESCRIPTION
**Pull request by yuvipanda**
Writing and reviewing terraform code has different challenges than
reviewing other kinds of code. This is a draft attempt at
codifying how code review and merges work with terraform in our
repo, so we can make changes safely and quickly

closes https://github.com/2i2c-org/team-compass/issues/213

### CODE REVIEW

**Review by sgibson91 — APPROVED** [review]
As a draft, I think this is a great start. Clearly explains why the code review is complex and how the different scenarios of code review differ with clear process for each 👍🏻 Thank you!

File: docs/contributing/code-review.md
Comment (damianavila) [inline docs/contributing/code-review.md:7]: So the idea of this main section is to add some general guidelines and then we go with specifics such as the terraform-related stuff, right?

File: docs/contributing/code-review.md
Comment (damianavila) [inline docs/contributing/code-review.md:25]: I like this section explaining the associated complexities!

**Review by damianavila — APPROVED** [review]
I left a question but nothing blocking this PR.
Even as a draft it is a great first step and I generally agree with the described workflows!

File: docs/contributing/code-review.md
Comment (yuvipanda) [reply docs/contributing/code-review.md:7]: Yep! I'm hoping we can add other guidelines soon. At the least, we should have something for:

1. doc changes
2. New hub creations
3. Hub config changes
4. ???

File: docs/contributing/code-review.md
Comment (yuvipanda) [reply docs/contributing/code-review.md:25]: yay!

File: docs/contributing/code-review.md
Comment (damianavila) [reply docs/contributing/code-review.md:7]: Those additional sections make sense to me. Not sure what else could be different from the general guidelines to deserve a specific section.

File: docs/contributing/code-review.md
Comment (choldgraf) [inline docs/contributing/code-review.md:34]: Can we add a step or checkbox like "explicitly mark when `terraform apply` has been run"?

File: docs/contributing/code-review.md
Comment (choldgraf) [inline docs/contributing/code-review.md:55]: Given that the person runs `terraform apply` on their own, what does the merge step do? Simply make the codebase reflect what has been applied? If so, then (related to my previous comment), there will be an in-between time when the state of the cluster is different from the state of `main`. What can we do to ensure that this difference is very short, or at least that it is clear to people?

**Review by choldgraf — APPROVED** [review]
This is helpful context on Terraform + I like the process around it. Thanks for this write-up!

I have two quick comments but they're both related to the same idea around "state of cluster" vs. "state of codebase". Curious what others think, but I don't think either should block this PR

File: docs/contributing/code-review.md
Comment (damianavila) [reply docs/contributing/code-review.md:34]: Makes sense to me!

File: docs/contributing/code-review.md
Comment (damianavila) [reply docs/contributing/code-review.md:55]: I don't think there is a way to enforce/ensure that... so we need policy. Maybe we need to be more strong and establish a max amount of time the difference could live??

Comment (sgibson91) [reply docs/contributing/code-review.md:55]: I agree that this is difficult to enforce, but I think `terraform apply` should be run immediately after the merge, by whomever merged.

Comment (yuvipanda) [reply docs/contributing/code-review.md:55]: > but I think terraform apply should be run immediately after the merge, by whomever merged.

The options are wether it is run before merge or after merge. I personally prefer before merge, since it allows for iteration when `terraform plan` fails (as it often does). If we were to run it after merge, it could break and we'll end up with terraform code in our main branch that fails...

**Comment by choldgraf** [comment]
I requested reviews from some more team members, since my push changes and updates the wording a little bit. It would be good if at least one person could give this a 👍 and make sure that we haven't missed anything important.

Also I can't request a review from @yuvipanda since he's the author, but I would love if he could confirm that this PR still looks good to him w/ my changes :-)

File: docs/contributing/code-review.md
Comment (consideRatio) [inline docs/contributing/code-review.md:74]: Are these lines to be deleted in favor of the `To deploy changes, follow these steps:` section just below?

```suggestion
```

File: docs/contributing/code-review.md
Comment (consideRatio) [inline docs/contributing/code-review.md:92]: I made a pass on the last four root-level bullet points. Feel free to edit my suggestion with tweaks to it as well. I wanted to make us note that we are both starting and stopping a deployment, as well as a clarification on what to do if you needed to make a changes to make it deploy successfully or not.

```suggestion
1. **Deploy the change**. Once approved, leave a comment in the PR that you will start deploying the changes. Then `terraform apply` your change and verify it works ok.
1. **Iterate as needed**. If things break or you need to iterate on the deployment, follow this cycle until the infrastructure is in the state you wish:

   1. Amend your PR with changes
   2. Deploy with `terraform apply`
   3. Comment that you've made a deployment
1. **Communicate that you've deployed**. Leave a comment in the PR that the changes have been deployed. You should always communicate when a deploy has been made!
1. **Self-merge or request another review**. If you didn't need to make a change to the PR, merge the PR. If you made changes to the PR, use your judgement to decide if you should request another review or if you should self-merge it directly and just leaving a comment about the change since it was initially approved.
```

**Review by consideRatio — COMMENTED** [review]
I made two suggestions, but otherwise this LGTM! Nice work!

**Comment by choldgraf** [comment]
many thanks for the review @consideRatio - changes accepted!

for others - let's leave this open for another 48 hours. If somebody else approves then I think we can just merge it in unless somebody objects.

**Review by damianavila — APPROVED** [review]
LGTM

**Comment by damianavila** [comment]
> LGTM

Not sure about the CI failure but content LGTM...

File: docs/contributing/code-review.md
Comment (choldgraf) [inline docs/contributing/code-review.md:87]:
```suggestion
```

**Comment by choldgraf** [comment]
OK - we've got a few approves, and CI is now happy, and we've given it several days, so I'm gonna merge this one in! Thanks for all the feedback everybody :-)

**Pull request merged**
